File: problems/20. Valid Parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        lookup = {")": "(", "]": "[", "}": "{"}
        stack = []
        for char in s:
            if char in "({[":
                stack.append(char)
            else:
                if not stack:
                    return False
                opening = stack.pop()
                if lookup[char] != opening:
                    return False

        return False if stack else True

File: problems/20. Valid Parentheses/solution.cpp
#include <unordered_map>
#include <stack>
#include <vector>
class Solution
{
public:
    bool isValid(std::string s)
    {
        std::unordered_map<char, char> lookup{{')', '('}, {']', '['}, {'}', '{'}};
        std::stack<char> stack;

        for (char c : s)
        {
            std::string openBrackets = "({[";
            if (openBrackets.find(c) != std::string::npos)
            {
                stack.push(c);
            }
            else
            {
                if (stack.size() == 0)
                    return false;
                char opening = stack.top();
                stack.pop();
                if (opening != lookup[c])
                    return false;
            }
        }

        return stack.size() == 0;
    }
};